app/src/ui/diff/side-by-side-diff.tsx

```diff
@@ -408,6 +411,89 @@ export class SideBySideDiff extends React.Component<
         }
       }
     }
+
+    if (this.state.lastExpandedHunk !== prevState.lastExpandedHunk) {
+      this.focusAfterLastExpandedHunkChange()
+    }
+  }
+
+  /**
+   * This handles app focus after a user has clicked on an diff expansion
+   * button. With the exception of the top expand up button, the expansion
+   * buttons disappear after clicking and by default the focus moves to the app
+   * body. This is not ideal for accessibilty as a keyboard user must then tab
+   * all the way back to the diff to continut to interact with it.
+   *
+   * If an expansion button of the type clicked is available, we focus it.
+   * Otherwise, we try to find the next closest expansion button and focus that.
+   * If no expansion buttons available, we focus the diff container. This makes
+   * it so if a user expands down and can expand down further, they will
+   * automatically be focused on the next expand down.
+   *
+   * Other context:
+   * - When a user clicks on a diff expansion button, the
+   * lastExpandedHunk state is updated. In the componentDidUpdate, we detect
+   * that change in order to call this after the new expansion buttons have
+   * rendered. The rendered expansion buttons are stored in a map.
+   * - A hunk index may have multiple expansion buttons (up and down) so it does
+   *   not uniquely identify a button.
+   */
+  private focusAfterLastExpandedHunkChange() {
+    if (this.state.lastExpandedHunk === null) {
+      return
+    }
+
+    const diffNode = findDOMNode(this.virtualListRef.current)
+    const diff = diffNode instanceof HTMLElement ? diffNode : null
+
+    // No expansion buttons? Focus the diff
+    if (this.hunkExpansionRefs.size === 0) {
+      diff?.focus()
+      return
+    }
+
+    const expansionHunkKeys = Array.from(this.hunkExpansionRefs.keys())
+    const { hunkIndex, expansionType } = this.state.lastExpandedHunk
+    const lastExpandedKey = `${hunkIndex}-${expansionType}`
+
+    // If there is a new hunk expansion button of same type in same place as the
+    // last, focus it
+    const lastExpandedHunkButton = this.hunkExpansionRefs.get(lastExpandedKey)
+    if (lastExpandedHunkButton) {
+      lastExpandedHunkButton.focus()
+      return
+    }
+
+    function getHunkKeyIndex(key: string) {
+      return parseInt(key.split('-').at(0) || '', 10)
+    }
+
+    // No?, Then try to focus the next closest hunk in tab order
+    const closestInTabOrder = expansionHunkKeys.find(
+      key => getHunkKeyIndex(key) >= hunkIndex
+    )
+
+    if (closestInTabOrder) {
+      const closetHunkButton = this.hunkExpansionRefs.get(closestInTabOrder)
+      closetHunkButton?.focus()
+      return
+    }
+
+    // No? Then try to focus the next closest hunk in reverse tab order
+    const closestInReverseTabOrder = expansionHunkKeys
+      .reverse()
```

Comment:
This seems to imply that `expansionHunkKeys` is already sorted by tab order but AFAICT it's sorted on insertion order (i.e. when the elements were added to the DOM) meaning it'll be sorted in tab order when scrolling downward but in the reverse order when scrolling upwards (since the list is virtualized). So we should probably sort `expansionHunkKeys` up front to ensure consistency regardless of scroll direction.

I'd also prefer if we, instead of reversing in-place here, used the [`findLast` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findLast) though that will require us adding `ES2023.Array` and I believe we'll have to land the TS5 upgrade first so depending on the order of things happening here we can keep .reverse or switch to .findLast

Reply:
Add the sort, `findLast` sound quite nice.